src/witness_gen.rs

```diff
@@ -0,0 +1,48 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use itertools::Itertools;
+use trustfall::FieldValue;
+use trustfall_rustdoc::VersionedRustdocAdapter;
+
+use crate::query::WitnessQuery;
+
+/// Runs the witness query of a given [`WitnessQuery`] for each lint query match, and merges the witness query
+/// results with the existing lint results. Each query must match exactly once, and will fail with an
+/// [`anyhow::Error`] otherwise.
+///
+/// Overlapping output keys between the [`WitnessQuery`] and the [`SemverQuery`](crate::query::SemverQuery)
+/// will result in the result from the [`WitnessQuery`] silently overriding the same key from the
+/// [`SemverQuery`](crate::query::SemverQuery).
+pub(crate) fn run_witness_queries(
+    adapter: &VersionedRustdocAdapter,
+    witness_query: &WitnessQuery,
+    lint_results: &Vec<BTreeMap<Arc<str>, FieldValue>>,
+) -> Vec<anyhow::Result<BTreeMap<Arc<str>, FieldValue>>> {
```

Comment:
Do we actually need this "`Vec` in `Vec` out" interface? I'm not sure what we gain by having this extra layer of abstraction, compared to just having a function that takes one lint result and produces the corresponding witness query output (or an error).